### PR TITLE
chasing 'pg: deadlock' issues when running integration tests

### LINKED ISSUE #631: Fix more random test failures
Some tests fail locally and on [Jenkins](https://ci.centos.org/job/devtools-fabric8-auth/1608/consoleFull) with the following error:

````
--- FAIL: TestCollaboratorsController (1.64s)
    --- FAIL: TestCollaboratorsController/TestListCollaborators (1.21s)
        --- FAIL: TestCollaboratorsController/TestListCollaborators/ok (1.19s)
            --- FAIL: TestCollaboratorsController/TestListCollaborators/ok/default (1.01s)
            	assertions.go:247: 
	Error Trace:	space_wrapper.go:45
			test_graph.go:76
			test_graph.go:98
			collaborators_blackbox_test.go:93
	Error:      	Received unexpected error:
	            	pq: deadlock detected
	            	github.com/fabric8-services/fabric8-auth/authorization/role/repository.(*GormRoleMappingRepository).Create
	            		/tmp/go/src/github.com/fabric8-services/fabric8-auth/authorization/role/repository/role_mapping.go:111
	            	github.com/fabric8-services/fabric8-auth/authorization/resource/service.(*resourceServiceImpl).Register.func1
	            		/tmp/go/src/github.com/fabric8-services/fabric8-auth/authorization/resource/service/resource_service.go:175
	            	github.com/fabric8-services/fabric8-auth/application/service/factory.(*serviceContextImpl).ExecuteInTransaction.func1.1
	            		/tmp/go/src/github.com/fabric8-services/fabric8-auth/application/service/factory/service_factory.go:90
	            	runtime.goexit
	            		/usr/lib/golang/src/runtime/asm_amd64.s:2337
	            	github.com/fabric8-services/fabric8-auth/application/service/factory.(*serviceContextImpl).ExecuteInTransaction.func1
	            		/tmp/go/src/github.com/fabric8-services/fabric8-auth/application/service/factory/service_factory.go:101
	            	github.com/fabric8-services/fabric8-auth/application/service/factory.(*serviceContextImpl).ExecuteInTransaction
	            		/tmp/go/src/github.com/fabric8-services/fabric8-auth/application/service/factory/service_factory.go:113
	            	github.com/fabric8-services/fabric8-auth/authorization/resource/service.(*resourceServiceImpl).Register
	            		/tmp/go/src/github.com/fabric8-services/fabric8-auth/authorization/resource/service/resource_service.go:121            		
	            	github.com/fabric8-services/fabric8-auth/test/graph.newSpaceWrapper
	            		/tmp/go/src/github.com/fabric8-services/fabric8-auth/test/graph/space_wrapper.go:44
	            	github.com/fabric8-services/fabric8-auth/test/graph.(*TestGraph).createAndRegister
	            		/tmp/go/src/github.com/fabric8-services/fabric8-auth/test/graph/test_graph.go:76
	            	github.com/fabric8-services/fabric8-auth/test/graph.(*TestGraph).CreateSpace
	            		/tmp/go/src/github.com/fabric8-services/fabric8-auth/test/graph/test_graph.go:98
	            	github.com/fabric8-services/fabric8-auth/controller_test.(*CollaboratorsControllerTestSuite).TestListCollaborators.func1.1
	            		/tmp/go/src/github.com/fabric8-services/fabric8-auth/controller/collaborators_blackbox_test.go:93
	            	testing.tRunner
	            		/usr/lib/golang/src/testing/testing.go:746
	            	runtime.goexit
	            		/usr/lib/golang/src/runtime/asm_amd64.s:2337
	Test:       	TestCollaboratorsController/TestListCollaborators/ok/default
            --- FAIL: TestCollaboratorsController/TestListCollaborators/ok/private_email (0.02s)
            	space_testing.go:285: invalid response status code: got 500, expected 200
````

### PR DESCRIPTION
**Pull request by xcoulon**
using a testgraph attached to the current test instance `t`

fixes #631

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

**Comment by fabric8cd** [comment]
@xcoulon snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-636-1`

File: controller/collaborators_blackbox_test.go [hunk 1]
```diff
@@ -406,88 +445,116 @@ func (s *CollaboratorsControllerTestSuite) TestRemoveSingleCollaborator() {
...
 
-		// when
-		payload := &app.AddManyCollaboratorsPayload{Data: []*app.UpdateUserID{{ID: g.CreateUser().IdentityID().String(), Type: idnType}}}
-		test.AddCollaboratorsForbidden(t, svc.Context, svc, ctrl, spaceID, g.CreateUser().IdentityID().String())
-		test.AddManyCollaboratorsForbidden(t, svc.Context, svc, ctrl, spaceID, payload)
-		rPayload := &app.RemoveManyCollaboratorsPayload{Data: []*app.UpdateUserID{{ID: toRemoveIdentity.ID.String(), Type: idnType}}}
-		test.RemoveManyCollaboratorsForbidden(t, svc.Context, svc, ctrl, spaceID, rPayload)
+		t.Run("add", func(t *testing.T) {
+			// when
+			payload := newAddManyCollaboratorsPayload(t, g.CreateUser().Identity())
```
Comment (sbryzak) [inline controller/collaborators_blackbox_test.go]: No need to create a new TestGraph here?

File: controller/collaborators_blackbox_test.go [hunk 2]
```diff
@@ -406,88 +445,116 @@ func (s *CollaboratorsControllerTestSuite) TestRemoveSingleCollaborator() {
...
+			// then
+			test.AddCollaboratorsForbidden(t, svc.Context, svc, ctrl, spaceID, g.CreateUser().IdentityID().String())
+			test.AddManyCollaboratorsForbidden(t, svc.Context, svc, ctrl, spaceID, payload)
+		})
+
+		t.Run("remove", func(t *testing.T) {
+			// when
+			payload := newRemoveManyCollaboratorsPayload(t, g.CreateUser().Identity())
```
Comment (sbryzak) [inline controller/collaborators_blackbox_test.go]: Or here also?

File: controller/collaborators_blackbox_test.go (hunk 1 above)
Comment (xcoulon) [reply controller/collaborators_blackbox_test.go]: the thing is, the data is not removed after each subtest, so here, we create a data set and check in the 2 subsequent subtests that operations are forbidden. I reckon that we could do the same (less data init) on other tests as well

File: controller/collaborators_blackbox_test.go (hunk 2 above)
Comment (xcoulon) [reply controller/collaborators_blackbox_test.go]: see response above

**Comment by fabric8cd** [comment]
@xcoulon snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-636-2`

File: controller/collaborators_blackbox_test.go (hunk 1 above)
Comment (alexeykazakov) [reply controller/collaborators_blackbox_test.go]: This is where you start make even deeper not-explicit dependencies between subtests which will make maintenance harder.
But I know that we are in disagreement here. You really like subtests and I would trade them for better isolation and auto-configuration/cleanup between tests ;)

Comment (xcoulon) [reply controller/collaborators_blackbox_test.go]: @alexeykazakov what dependency are you talking about ? if it's about sharing the same data set between those subtests, then I have absolutely no problem with moving/duplicating the `//given` section (ie, data init) within each subtest.
About isolation, I think that Ginkgo is better at that. See the `Extracting common setup` on http://onsi.github.io/ginkgo/#structuring-your-specs

Comment (xcoulon) [reply controller/collaborators_blackbox_test.go]: @alexeykazakov I've duplicated the data init in each subtest in https://github.com/fabric8-services/fabric8-auth/pull/636/commits/d22a6464812250925d924622caadb142ee22b9f6

**Comment by fabric8cd** [comment]
@xcoulon snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-636-3`

**Comment by fabric8cd** [comment]
@xcoulon snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-636-4`

**Comment by fabric8cd** [comment]
@xcoulon snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-636-5`

**Comment by fabric8cd** [comment]
@xcoulon snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-636-6`

**Pull request merged**